torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -235,30 +305,68 @@ void ProcessGroupGloo::createAlgorithm(AlgorithmEntry& entry) {
 template <typename T>
 void ProcessGroupGloo::createAllreduce(AlgorithmEntry& entry) {
   const auto& key = entry.key;
+  const auto& backend = key.type->backend();
 
   // Create algorithm against first context
   auto& context = contexts_[0];
-  entry.algorithm = std::unique_ptr<::gloo::Algorithm>(
-      new ::gloo::AllreduceHalvingDoubling<T>(
-          context,
-          getDataPointers<T>(entry.src),
-          entry.src[0].numel(),
-          reductionFunction<T>(key.reduceOp)));
+
+  if (backend == at::kCPU) {
+    entry.algorithm = std::unique_ptr<::gloo::Algorithm>(
+        new ::gloo::AllreduceHalvingDoubling<T>(
+            context,
+            getDataPointers<T>(entry.src),
+            entry.src[0].numel(),
+            reductionFunction<T>(key.reduceOp)));
+    return;
+  }
+
+  if (backend == at::kCUDA) {
+    entry.algorithm = std::unique_ptr<::gloo::Algorithm>(
```

Comment:
It would be nice to support GPU direct when creating the algorithm and when IB is available.  Can you create a TODO to automatically enable GPU direct when the device supports it?

Reply:
Added #8032